lab/lab-guide.md

```diff
@@ -0,0 +1,802 @@
+
+    ```csharp
+    //string prompts = string.Empty;
+    //float[] vectors = new float[1536];
+    //string response = string.Empty;
+    //await Task.Delay(0);
+    ```
+
+1. Next add the following lines below the commented code (the return will already be there). The function should look like this below.
+
+    ```csharp
+    private async Task<(string cachePrompts, float[] cacheVectors, string cacheResponse)> GetCacheAsync(List<Message> contextWindow)
+    {
+     //string prompts = string.Empty;
+     //float[] vectors = new float[1536];
+     //string response = string.Empty;
+     //await Task.Delay(0);
+      string prompts = string.Join(Environment.NewLine, contextWindow.Select(m => m.Prompt));
+      float[] vectors = await _semanticKernelService.GetEmbeddingsAsync(prompts);
+      string response = await _cosmosDbService.CacheGetAsync(vectors, _cacheSimilarityScore);
+      
+      return (prompts, vectors, response);
+    }
+     ```
+
+1. Save the file, and optionally, delete the commented code from the function.
+
+1. We will now implement the vector search for our cache. Locate the the **Services/CosmosDbService.cs** class. Then look for a function called, **CacheGetAsync()** with the signature below.
+
+    ```csharp
+    public async Task<string> CacheGetAsync(float[] vectors, double similarityScore)
+    ```
+    
+1. Next, comment out the empty **queryText** and replace it with the vector search query. This query is what performs the search to find relevant items in our semantic cache. It selects only the closest match with a descending *ORDER BY*, so the item with highest similarity score (and the most similar to what is being searched) appears first. The results include a previously cached completion, and the similarity score for the cached item.
+
+    ```csharp
+    //string queryText = @"";
+    string queryText = $"""
+    SELECT Top 1 
+        c.prompt, c.completion, VectorDistance(c.vectors, @vectors) as similarityScore
+    FROM c  
+    WHERE 
+        VectorDistance(c.vectors, @vectors) > @similarityScore 
+    ORDER BY 
+        VectorDistance(c.vectors, @vectors)
+    """;
+    ```
+
+1. Save the file to save the new query text.
+
+1. Return to the **ChatService.cs** and find the **GetChatCompletionAsync()** function with the following signature. We will modify this function to search our semantic cache for user prompts with a similar intent to what is being asked. If we find a match, we can return the cached value to the user instead of making a new request to the LLM.
+
+    ```csharp
+    public async Task<Message> GetChatCompletionAsync(string tenantId, string userId, string? sessionId, string promptText)
+    ```
+
+1. Comment out the below line in **ChatService.cs** inside the **GetChatCompletionAsync()** function.
+
+    ```csharp
+    //(chatMessage.Completion, chatMessage.CompletionTokens) = await _semanticKernelService.GetRagCompletionAsync(sessionId, contextWindow, promptText, _productMaxResults);
```

Comment:
btw, I removed sessionId from this function signature since it's no longer being used.